Mathlib/Tactic/CasesM.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2022 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro
+-/
+import Lean
+
+/-!
+# `casesm`, `cases_type`, `constructorm` tactics
+
+These tactics implement repeated `cases` / `constructor` on anything satisfying a predicate.
+-/
+
+namespace Mathlib.Tactic
+open Lean Meta Elab Tactic
+
+/--
+Core tactic for `casesm` and `cases_type`. Calls `cases` on all fvars in `g` for which
+`matcher ldecl.type` returns true.
+* `recursive`: if true, it calls itself repeatedly on the resulting subgoals
+* `allowSplit`: if false, it will skip any hypotheses where `cases` returns more than one subgoal.
+-/
+partial def casesMatching (g : MVarId) (matcher : Expr → MetaM Bool)
+    (recursive := false) (allowSplit := true) : MetaM (List MVarId) := return (← go g).toList where
+  /-- Auxiliary for `casesMatching`. Accumulates generated subgoals in `acc`. -/
+  go (g : MVarId) (acc : Array MVarId := #[]) : MetaM (Array MVarId) :=
+    g.withContext do
+      for ldecl in ← getLCtx do
+        if ← matcher ldecl.type then
+          let mut acc := acc
+          let subgoals ← if allowSplit then
+            g.cases ldecl.fvarId
+          else
+            let s ← saveState
+            let subgoals ← g.cases ldecl.fvarId
+            if subgoals.size > 1 then
```

Comment:
I think you're missing an `if allowSplit` here.

Reply:
this is in the `else` branch of an `if allowSplit`